### DISCUSSION TITLE
Crash after applying close_merge_vertices && Ball pivoting

### DISCUSSION THREAD

**Discussion by wikyburns** [cnr-isti-vclab/PyMeshLab#96] [Q&A] [answered]
[testing_file.zip](https://github.com/cnr-isti-vclab/PyMeshLab/files/6285273/testing_file.zip)

Hi, I have a point cloud, and when I apply the filter: merge_close_vertices it removes everything.

I tested that on UI Meshlab and it works great but not with the pymeshlab.

ms = pymeshlab.MeshSet()
ms.load_new_mesh('toConvert/'+file_name)
ms.apply_filter('merge_close_vertices', threshold=12)
ms.apply_filter('surface_reconstruction_ball_pivoting', ballradius= 0, clustering=20, creasethr= 90, deletefaces=0)
ms.apply_filter('laplacian_smooth', stepsmoothnum=3, selected=0)
ms.apply_filter('transform_scale_normalize', axisx=1000, axisy=1000, axisz=1000, uniformflag=1, scalecenter='origin', customcenter=[0,0,0], unitflag=0, freeze=1, alllayers=0)
ms.apply_filter('transform_translate_center_set_origin', traslmethod='Center on Scene BBox')
ms.save_current_mesh(fileName+'.obj')
ms.save_current_mesh(fileName+'.stl')

Thanks,

**Comment by alemuntoni** [comment, marked as answer]
`ms.apply_filter('merge_close_vertices', threshold=12)` treats the parameter `threshold` as a float (absolute value).
To treat it as a percentage, you need to pass an object of type [Percentage](https://pymeshlab.readthedocs.io/en/latest/classes/percentage.html) to the parameter `threshold`.

Take a look to [this tutorial](https://pymeshlab.readthedocs.io/en/latest/tutorials/apply_filter_parameters_percentage.html).

**Reply by wikyburns** [reply]
Hi @alemuntoni , I tried with the object Percentage and now it's doing nothing, just stopping with any type of error, I tried applying the filter exported from meshlab and the same:

```
percentageMerge = pymeshlab.Percentage(12.00)
ms = pymeshlab.MeshSet()
ms.load_new_mesh('toConvert/'+file_name)
ms.apply_filter('merge_close_vertices', threshold=percentageMerge)
#At this point returns {}
```

`ms.apply_filter('surface_reconstruction_ball_pivoting', ballradius= 0, clustering=20, creasethr= 90, deletefaces=0)`
After apply this filter the script just stops without any error.

Thanks,

**Reply by wikyburns** [reply]
Good evening @alemuntoni do you have any idea about how I can fix it?

Really thanks for the help

Regards

**Reply by alemuntoni** [reply]
> Hi @alemuntoni , I tried with the object Percentage and now it's doing nothing, just stopping with any type of error, I tried applying the filter exported from meshlab and the same:
> 
> ```
> percentageMerge = pymeshlab.Percentage(12.00)
> ms = pymeshlab.MeshSet()
> ms.load_new_mesh('toConvert/'+file_name)
> ms.apply_filter('merge_close_vertices', threshold=percentageMerge)
> #At this point returns {}
> ```
> 
> `ms.apply_filter('surface_reconstruction_ball_pivoting', ballradius= 0, clustering=20, creasethr= 90, deletefaces=0)`
> After apply this filter the script just stops without any error.
> 
> Thanks,

Hi, 

I just don't understand what the issue is here. If it stops without any type of error, maybe everything was successful??
If there is no crash or exception thrown, it means that the filter has been applied correctly.

**Reply by wikyburns** [reply]
Hi @alemuntoni 

Can you try it on your server if you wish, it just closes with any type of result, cause python just stops on try to apply the filter: surface_reconstruction_ball_pivoting.

![image](https://user-images.githubusercontent.com/11525781/115604905-2dd25600-a2e2-11eb-90fe-743638aa3234.png)


any idea? on the UI meshlab works perfectly.

Really thanks man!

**Reply by alemuntoni** [reply]
oh ok! that's strange, it seems that it is crashing... I'll check on this.

**Reply by alemuntoni** [reply]
fixed [here](https://github.com/cnr-isti-vclab/PyMeshLab/commit/22b0c9661b95abd7f8c33b17cab8fb13aa666ee9), the issue was in the ball pivoting filter in meshlab.
It will be available in the next version of pymeshlab. In the meantime, as a workaround, before calling the ball pivoting filter, you can call:
```
ms.current_mesh().compact()
```

**Reply by wikyburns** [reply]
> fixed [here](https://github.com/cnr-isti-vclab/PyMeshLab/commit/22b0c9661b95abd7f8c33b17cab8fb13aa666ee9), the issue was in the ball pivoting filter in meshlab.
> 
> It will be available in the next version of pymeshlab. In the meantime, as a workaround, before calling the ball pivoting filter, you can call:
> 
> ```
> 
> ms.current_mesh().compact()
> 
> ``` 

@alemuntoni really nice! 

Great job! 

Thanks!! 👏🏼👏🏼👏🏼
